### PR TITLE
Cache RetargetingParameterSymbol.TypeWithAnnotations to avoid ~8.3% allocations in a trace

### PR DESCRIPTION
**Pull request by Youssef1313**
In the trace from https://developercommunity.visualstudio.com/t/Performance-trace-for-Roslyn-team/10413054, there are too many `TypeWithAnnotations[]` allocations happening in RetargetingParameterSymbol (thanks @sharwell for the investigation).

image

I re-profiled the scenario with a locally built version of the compiler that adds some logging to CodeAnalysisEventSource, and @sharwell figured out that on average, there are 76 calls to the exact same `RetargetingParameterSymbol` instance. So we think caching might be beneficial here.

I don't know if that will solve the problem completely, we might need to get another trace with the change from this PR and see if there are more bottlenecks, but at least, this seems like a good first step.

Edit:

The previous screenshot from trace shows 3.5% allocations for TypewithAnnotions arrays, however, there is **also** the following in the same trace:

image

which is 4.8% of allocations. So total is 8.3%, and there are probably more allocations happening during the evaluation of this property.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs
Comment (cston) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:45]: `Nullable<TypeWithAnnotations>` is not thread-safe. Is this field accessed from multiple threads?

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:45]: @cston Yeah I was chatting with Sam about this. I didn't find an appropriate existing helper, so Sam is working on merging helpers that are not available in the compilers layer with those available in the compilers layer.

Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:45]: No additional helpers are necessary, that's what https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/TypeWithAnnotations.cs,e5078e3d5fd03b05,references is for.

Comment (cston) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:45]: If the goal is to avoid recalculating the type, it may be sufficient to just cache the `TypeSymbol`.
```csharp
private TypeSymbol? _lazyType;

public sealed override TypeWithAnnotations TypeWithAnnotations
{
    get
    {
        var underlyingType = _underlyingParameter.TypeWithAnnotations;
        var translator = RetargetingModule.RetargetingTranslator;
        if (_lazyType is null)
        {
            Interlocked.CompareExchange(ref _lazyType,
                translator.Retarget(underlyingType.Type,
                    RetargetOptions.RetargetPrimitiveTypesByTypeCode),
                null);
        }
        var newModifiers = translator.RetargetModifiers(underlyingType.CustomModifiers, out _);
        return underlyingType.WithTypeAndModifiers(_lazyType, newModifiers);
    }
}
```

**Comment by 333fred** [comment]
I would like to better understand the scenario causing an issue here. Can you share the code being compiled?

**Comment by Youssef1313** [comment]
@333fred The code I'm currently testing with is a private repo. I don't think I'll be able to share it. But I think the repro from https://github.com/dotnet/roslyn/issues/67926 will do it. It's almost the same idea. The PR https://github.com/dotnet/roslyn/pull/68316/ did have a good impact on this scenario, but it doesn't resolve it completely.

It's essentially a codebase full of generic extension methods. In a 3 minute trace, `RetargetingParameterSymbol.TypeWithAnnotations` was accessed over 10 million times, to be accurate, 12,833,732.

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs
Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:45]: @cston In the trace I provided, this property is accessed 12,833,732 times during 3 minutes. So I think it's worth to just cache the whole thing.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:45]: > No additional helpers are necessary, that's what https://sourceroslyn.io/#Microsoft.CodeAnalysis.CSharp/Symbols/TypeWithAnnotations.cs,e5078e3d5fd03b05,references is for.

Does this suggest it like that?

```csharp
if (_lazyTypeWithAnnotations is null)
{
    Interlocked.CompareExchange(ref _lazyTypeWithAnnotations, new TypeWithAnnotations.Boxed(......), null);
}
```

I think that would still risk calculating it twice if two threads concurrently accessed the property.

What Sam suggested was something to be like this:

```csharp
return LazyInitialization.EnsureInitialized(
    ref _lazyTypeWithAnnotations,
    static self => self.RetargetingModule.RetargetingTranslator.Retarget(self._underlyingParameter.TypeWithAnnotations, RetargetOptions.RetargetPrimitiveTypesByTypeCode),
    this);
```

**Comment by Youssef1313** [comment]
@333fred In case you wanted me to log more events into the trace, just let me know.

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:26]: >private TypeWithAnnotations? _lazyTypeWithAnnotations; [](http://example.com/codeflow?start=8&length=54)

We use ```TypeWithAnnotations.Boxed``` to cache this information in symbols. There should be many examples

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:45]: @333fred The helpers are lazy initialization helpers for reference types, and are unrelated to the fact that this is a value type.

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs
Comment (sharwell) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:26]: ➡️ This is a duplicate of the conversation in https://github.com/dotnet/roslyn/pull/69011#discussion_r1262022836, so we'll continue the discussion there.

**Comment by sharwell** [comment]
In addition to the number of times `TypeWithAnnotations` is accessed, we showed that _on average_ the `TypeWithAnnotations` property is accessed 76 times for each `RetargetingParameterSymbol` instance.

**Comment by sharwell** [comment]
Blocked on #69017

**Comment by 333fred** [comment]
> In a 3 minute trace, RetargetingParameterSymbol.TypeWithAnnotations was accessed over 10 million times, to be accurate, 12,833,732.

I want to understand the codepath that caused this. This seems extreme and like some component is misbehaving. I'm sympathetic to the issue, but I don't want to patch over the problem, I want to understand why it is a problem.

**Comment by Youssef1313** [comment]
@333fred Does the stacks in PerfView not give enough info about the code path causing this?

File: src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Retargeting/RetargetingParameterSymbol.cs:45]: We should switch to ```TypeWithAnnotations.Boxed```, which we use to cache this information in symbols. There should be many examples. No special helpers are necessary and the change shouldn't be blocked on that.

**Comment by AlekseyTs** [comment]
> Blocked on #69017

Why is that? Regular ```Interlocked.CompareExchange``` just works with ```TypeWithAnnotations.Boxed```.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 2)

**Comment by AlekseyTs** [comment]
@dotnet/roslyn-compiler For the second review

**Comment by CyrusNajmabadi** [comment]
Do you have a trace showing the 'after' results?  Thanks!

**Comment by Youssef1313** [comment]
@CyrusNajmabadi Sure. I uploaded it to the same feedback ticket on DC.

**Review by sharwell — APPROVED** [review]

**Comment by Youssef1313** [comment]
@333fred Given the before/after traces, can this move forward and get merged?

**Review by 333fred — APPROVED** [review]
I'm satisfied by scenario here.

**Pull request merged**
